### PR TITLE
catalog: Optimize catalog initialization

### PR DESCRIPTION
**Pull request by jkosh44**
`TableTransaction`s provide a method called `set` that allows someone to set the value of a specific key. After setting the value, the method runs a uniqueness check on all keys and values. Certain code paths call `set` in a loop, causing the uniqueness check to be run multiple times.

This commit adds a new method to `TableTransaction` called `set_many`, that sets the value of many keys and then runs a single uniqueness check. This allows us to reduce latency in the cases where we were calling `set` in a loop.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
Currently, there isn't a huge speedup from this code, but I have an incoming change that makes heavy use of `set_many`.

File: src/stash/src/lib.rs [hunk 1]
```diff
@@ -825,6 +827,57 @@ where
         }
     }
 
+    /// Set the values for many keys. Returns the previous entry for each key if the key existed,
+    /// otherwise None.
+    ///
+    /// Returns an error if any uniqueness check failed.
+    pub fn set_many(
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/stash/src/lib.rs:834]: I like it as is, even with a bit of kind of code duplication. Avoiding allocs is nice.

This function looks good, but I'd like it to be tested within the stash crate too.

[the remark this answers is not in the document]
Comment (jkosh44) [reply src/stash/src/lib.rs:834]: Just added some tests next to the `set` tests.

**Comment by jkosh44** [comment]
Note: I've set this to auto-merge, so don't approve unless you're OK with merging as is.

**Review by mjibson — APPROVED** [review]

**Pull request merged**